sklearn/linear_model/tests/test_least_angle.py

```diff
@@ -686,3 +686,109 @@ def test_lasso_lars_vs_R_implementation():
 
     assert_array_almost_equal(r2, skl_betas2, decimal=12)
     ###########################################################################
+
+
+def test_lasso_lars_copyX_behaviour1():
+    """
+    Test that user input regading copyX is not being overridden (it was until
+    at least version 0.21)
+
+    Correct behaviour is not to create a copy.
+
+    """
+    temp = np.asarray
+
+    def ident(x, *args, **kwargs):
+        return x
+
+    np.asarray = ident
+    lasso_lars = LassoLarsIC(copy_X=False, precompute=False)
```

Comment:
can't you do something simpler. Check that values have changed or not? that's what we do in a number of other places in tests.

Reply:
It  would be great to do something simpler. Monkey-patching two numpy functions is indeed a bit heavy-handed. I tried a test based on creating a copy before `.fit` and comparing it to X after `fit`. TL;DR: it didn't work.  Long version: I couldn't make the current version fail (even though it should) because of the following: there are two places where copies are created inside `fit`, `_preprocess_data` and `lars_path`. This is where the current mixed behaviour arises. `_preprocess_data` uses `self._copy_X` (from `__init__`) and `lars_path` uses `copy_X', a parameter to 'fit' with default value of 'True'. If the user did not want a copy a set 'copy_X' to 'False' when creating the object, then `_preprocess_data` will not make a copy and modify the values in place. This already satisfies the test. However, in the next step, 'lars_path' will make a copy (with a potentially massive memory cost) but the test will pass because the values were modified.

Reply:
ok. I would still suggest the simpler route. Have test that checks the 3 options copy_X in in init or fit makes sure X passed in not modified. Else it is modified and then if we can make the still test pass while avoid yet another copy of the preprocessed X it is enough to show that the behavior is correct. ok?

Reply:
I think I understand, let me rephrase to be sure: make a test against the possibility of input data being modified in place when the user explicitly wanted the data preserved (clearly the most destructive behaviour); do not make a test against the possibility of data being copied when it was not necessary (because the user was ok with the data being modified). For these tests, it is enough to compare the data against changed values, and tapping into `__copy__` is not needed.